clippy_lints/src/methods/unnecessary_map_or.rs

```diff
@@ -109,10 +109,16 @@ pub(super) fn check<'a>(
         );
 
         let sugg = if let Some(parent_expr) = get_parent_expr(cx, expr) {
-            match parent_expr.kind {
-                ExprKind::Binary(..) | ExprKind::Unary(..) | ExprKind::Cast(..) => binop.maybe_paren(),
-                ExprKind::MethodCall(_, receiver, _, _) if receiver.hir_id == expr.hir_id => binop.maybe_paren(),
-                _ => binop,
+            if parent_expr.span.from_expansion() {
```

Comment:
IMO it would be better/more semantically correct to compare `SyntaxContext`s of the two spans here, like

```suggestion
            if !parent_expr.span.eq_ctxt(expr.span) {
```

Right now this won't really make a difference since `methods/mod.rs` ignores any expressions that are from macro expansions so `expr` is always from the root context, but I think it could make sense to eventually lint some local macro cases like this, in which case this condition wouldn't be correct:
```rs
macro_rules! m {
  ($x:expr) => {
    !$x.map_or(false, |v| v == 1)    // should become !($x == Some(1))
  }
}
```
We still need parentheses here if the lint fires on that code even though the parent negate expr is from macro code; comparing `SyntaxContext` remains correct here

Reply:
Makes sense! I'd like to add a test case for this, but I guess it's not really possible with the current logic of `methods/mod.rs` as you say? Should I still add the example you give, but commented out and with a FIXME saying to uncomment once possible?

Reply:
Yeah, adding a test case for future proofing the lint even if it doesn't emit there yet sounds like a good idea to me. It doesn't even necessarily be commented out IMO, it would be fine to just have the test case in the file that doesn't emit yet and has a FIXME that it would be desirable to lint like you said.

If you want to see if your code would handle this situation fine you could temporarily locally remove these two checks and see what the suggestion would be:
https://github.com/rust-lang/rust-clippy/blob/651ca3372b1068fa366147b313be56a9c7f10b23/clippy_lints/src/methods/mod.rs#L4776
https://github.com/rust-lang/rust-clippy/blob/651ca3372b1068fa366147b313be56a9c7f10b23/clippy_lints/src/methods/mod.rs#L4765

Reply:
I first needed to also remove this line for the lints to kick in: https://github.com/rust-lang/rust-clippy/blob/651ca3372b1068fa366147b313be56a9c7f10b23/clippy_lints/src/methods/mod.rs#L4766

I pasted the test case in, but it (understandably) didn't fire at that point, because `$x` is not yet known to be an `Option<{integer}>`. So I added an invocation like this:
```rs
macro_rules! m {
    ($x:expr) => {
        let _ = !$x.map_or(false, |v| v == 1);
        //~^ unnecessary_map_or
        let _ = $x.map_or(false, |v| v == 1);
        //~^ unnecessary_map_or
    };
}
m!(Some(5));
```

But on that, I got the following suggestions:
```
error: this `map_or` can be simplified
  --> tests/ui/u2.rs:6:22
   |
LL |             let _ = !$x.map_or(false, |v| v == 1);
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
LL |     m!(Some(5));
   |     ----------- in this macro invocation
   |
   = note: `-D clippy::unnecessary-map-or` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::unnecessary_map_or)]`
   = note: this error originates in the macro `m` (in Nightly builds, run with -Z macro-backtrace for more info)
help: use a standard comparison instead
   |
LL -             let _ = !$x.map_or(false, |v| v == 1);
LL +             let _ = !(Some(5) == Some(1));
   |

error: this `map_or` can be simplified
  --> tests/ui/u2.rs:8:21
   |
LL |             let _ = $x.map_or(false, |v| v == 1);
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
LL |     m!(Some(5));
   |     ----------- in this macro invocation
   |
   = note: this error originates in the macro `m` (in Nightly builds, run with -Z macro-backtrace for more info)
help: use a standard comparison instead
   |
LL -             let _ = $x.map_or(false, |v| v == 1);
LL +             let _ = Some(5) == Some(1);
   |
```
And I think it's incorrect to suggest replacing `$x` since that changes the meaning of the macro. Ideally, we could wish for the suggestions to look like this instead:
```rs
let _ = !($x == Some(1));
let _ = $x == Some(1);
```
But again, that would require the lint to speculate about the type of `$x`, which I'm not sure is feasible.

Additionally, another case started to lint:
https://github.com/rust-lang/rust-clippy/blob/651ca3372b1068fa366147b313be56a9c7f10b23/tests/ui/unnecessary_map_or.rs#L58

Namely, it now got the following suggestion:
```
error: this `map_or` can be simplified
  --> tests/ui/unnecessary_map_or.rs:58:13
   |
LL |     let _ = x!().map_or(false, |n| n == 1);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: use a standard comparison instead
   |
LL -     let _ = x!().map_or(false, |n| n == 1);
LL +     let _ = Some(1) == Some(1);
   |
```
which I'm not quite sure about, since it involves inlining the definition of `x!()`.

I guess this would also be okay if we were able to retain the pre-expansion source, and suggest `let _ = x!() == Some(1);` instead.

So it looks like there are some kinks when going across a macro boundary, but resolving them would probably be a part of the effort to let lints in `methods/mod.rs` work in macros, as you said, so should I leave the new test case in?

Reply:
Ah okay, well seems like the lint isn't very equipped to deal with macros yet anyway then. Fixing the `$x` being inlined in the suggestion involves changing `Sugg::hir_with_applicability` to `Sugg::hir_with_context`.

Either way though, this is beyond the scope of the PR. Yep, I'd still just leave the test in and ignore all the other pre-existing macro issues for now, it's okay if it doesn't fire there.

Although at least I can see in your snippet that it's getting the parentheses right.

Reply:
Done. But now I'm not sure whether the changelog message/PR description is still accurate. Do you have any suggestions?..

Reply:
I think the changelog message is fine. Well, it's correct from a user's point of view, and they are primarily the ones who read those.